# **Predicciones de Ventas Alimenticias**
## Erwin Barriga

In [308]:
import pandas as pd
import numpy as np

In [309]:
filename = "./sales_predictions.csv"

In [310]:
df = pd.read_csv(filename)
df.head()  

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

### ¿Cuántas filas y columnas?

In [311]:
df.shape

(8523, 12)

### ¿Cuáles son los tipos de datos en cada variable?

In [312]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


### ¿Hay duplicados?

In [313]:
df.duplicated().sum()

0

###  Identifiquen los valores faltantes.

In [314]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

#### Columna: Item Weight

In [315]:
# dataframe donde "Item_Weight" == NaN
df[(df["Item_Weight"].isna())]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

Tomando un item al azar, hay filas donde el peso de ese item ha sido registrado con un mismo valor

In [316]:
df[df["Item_Identifier"] == "FDU28"]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
9              FDU28         19.2          Regular         0.094450   
6362           FDU28          NaN          Regular         0.164439   
6419           FDU28         19.2          Regular         0.094301   
7597           FDU28          NaN          Regular         0.093464   
8419           FDU28         19.2          Regular         0.093918   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
9     Frozen Foods  187.8214            OUT017                       2007   
6362  Frozen Foods  188.4214            OUT019                       1985   
6419  Frozen Foods  187.5214            OUT018                       2009   
7597  Frozen Foods  189.1214            OUT027                       1985   
8419  Frozen Foods  189.8214            OUT046                       1997   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
9            NaN               Tier 2  Supermarket Type1          4710.5350  
6362       Small               Tier 1      Grocery Store           376.8428  
6419      Medium               Tier 3  Supermarket Type2          1695.7926  
7597      Medium               Tier 3  Supermarket Type3          2261.0568  
8419       Small               Tier 1  Supermarket Type1           942.1070

**Estrategia:**

Buscar todos los items que tienen peso NaN, usar los valores de peso registrado en otras filas para inferir el valor de peso que sustituya los NaN

In [317]:
#* Extraer los "Item_Identifier" donde "Item_Weight" == NaN
df_items = df[(df["Item_Weight"].isna())]["Item_Identifier"]
df_items

7       FDP10
18      DRI11
21      FDW12
23      FDC37
29      FDC14
        ...  
8485    DRK37
8487    DRG13
8488    NCN14
8490    FDU44
8504    NCN18
Name: Item_Identifier, Length: 1463, dtype: object

In [318]:
df_items.duplicated().sum()

321

In [319]:
df_items = df_items.drop_duplicates()
df_items.duplicated().sum()

0

In [320]:
# Generar lista python de Items cuyo peso sea NaN
nanItemList = df_items.to_list()
#print(nanItemList)


In [321]:
# Remplazar NaN por valor de peso registrado en otras filas
for item in nanItemList:
    # print()
    # print(df[df["Item_Identifier"] == item]["Item_Weight"])
    ps_peso_item = df[df["Item_Identifier"] == item]["Item_Weight"]
    # print(ps_peso_item)
    valor_reemplazo = ps_peso_item.mean()
    df.loc[(df["Item_Identifier"] == item) & (df["Item_Weight"].isna()), "Item_Weight"] = valor_reemplazo

Verificar cuantos NaN quedan en la columna "Item_Weight"

In [322]:
df["Item_Weight"].isna().sum()

4

In [323]:
df[df["Item_Weight"].isna()]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
927            FDN52          NaN          Regular         0.130933   
1922           FDK57          NaN          Low Fat         0.079904   
4187           FDE52          NaN          Regular         0.029742   
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
927   Frozen Foods   86.9198            OUT027                       1985   
1922   Snack Foods  120.0440            OUT027                       1985   
4187         Dairy   88.9514            OUT027                       1985   
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
927       Medium               Tier 3  Supermarket Type3          1569.9564  
1922      Medium               Tier 3  Supermarket Type3          4434.2280  
4187      Medium               Tier 3  Supermarket Type3          3453.5046  
5022       Small               Tier 1      Grocery Store           120.5098

4 items únicos cuyo peso no está registrado de un universo de 8523 -> Eliminamos esas cuatro filas

In [324]:
df.dropna(subset=['Item_Weight'], inplace=True)
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

#### Columna: Outlet_Size
Realizamos algunas consultas para establecer algún patrón que relacione los valores NaN de la columma "Outlet_Size" con otras columnas

In [325]:
# 3 tipos de valores posibles para la columna Size
df[["Outlet_Size"]].value_counts()

Outlet_Size
Medium         2790
Small          2387
High            932
dtype: int64

In [326]:
# Dataframe donde "Outlet_Size" == NaN
df[df["Outlet_Size"].isna()]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular         0.000000   
8              FDH17       16.200          Regular         0.016687   
9              FDU28       19.200          Regular         0.094450   
25             NCD06       13.000          Low Fat         0.099887   
28             FDE51        5.925          Regular         0.161467   
...              ...          ...              ...              ...   
8502           NCH43        8.420          Low Fat         0.070712   
8508           FDW31       11.350          Regular         0.043246   
8509           FDG45        8.100          Low Fat         0.214306   
8514           FDA01       15.000          Regular         0.054489   
8519           FDS36        8.380          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

In [327]:
# Los  Outlet_Identifier que tienen la columna Size = NaN, lo tienen en todo el registro
df[df["Outlet_Identifier"] == "OUT010"]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular         0.000000   
28             FDE51        5.925          Regular         0.161467   
30             FDV38       19.250          Low Fat         0.170349   
45             FDM39        6.420               LF         0.089499   
65             FDC46       17.700               LF         0.195068   
...              ...          ...              ...              ...   
8400           NCE07        8.180          Low Fat         0.021977   
8432           FDI16       14.000          Regular         0.227261   
8473           DRI47       14.700          Low Fat         0.035016   
8486           FDR20       20.000          Regular         0.000000   
8509           FDG45        8.100          Low Fat         0.214306   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
28                    Dairy   45.5086            OUT010   
30                    Dairy   55.7956            OUT010   
45                    Dairy  178.1002            OUT010   
65              Snack Foods  185.4266            OUT010   
...                     ...       ...               ...   
8400              Household  140.6154            OUT010   
8432           Frozen Foods   54.3640            OUT010   
8473            Hard Drinks  144.3128            OUT010   
8486  Fruits and Vegetables   46.4744            OUT010   
8509  Fruits and Vegetables  213.9902            OUT010   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
28                         1998         NaN               Tier 3   
30                         1998         NaN               Tier 3   
45                         1998         NaN               Tier 3   
65                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8400                       1998         NaN               Tier 3   
8432                       1998         NaN               Tier 3   
8473                       1998         NaN               Tier 3   
8486                       1998         NaN               Tier 3   
8509                       1998         NaN               Tier 3   

        Outlet_Type  Item_Outlet_Sales  
3     Grocery Store           732.3800  
28    Grocery Store           178.4344  
30    Grocery Store           163.7868  
45    Grocery Store           358.2004  
65    Grocery Store           184.4266  
...             ...                ...  
8400  Grocery Store           425.4462  
8432  Grocery Store           159.7920  
8473  Grocery Store           431.4384  
8486  Grocery Store            45.2744  
8509  Grocery Store           424.7804  

[555 rows x 12 columns]

In [328]:
#   Todos los Outlets que tienen la columna Size igual a NaN, agrupados por "Outlet_Type"
df[df["Outlet_Size"].isna()]["Outlet_Type"].value_counts()

Supermarket Type1    1855
Grocery Store         555
Name: Outlet_Type, dtype: int64

In [329]:
#   Todos los Outlet_Types que son Grocery Store, tienen la columna Size igual "Small"
df[(df["Outlet_Type"]=="Grocery Store")]["Outlet_Size"].value_counts()

Small    527
Name: Outlet_Size, dtype: int64

**Estrategia:**

Para todos los outlets cuyo tamaño es NaN y que sean del tipo "Grocery Store", usar el valor "Small" como sustituto.

In [330]:
df.loc[(df["Outlet_Type"]=="Grocery Store") & (df["Outlet_Size"].isna()), "Outlet_Size"] = "Small"
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  1855
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Se redujo el número de NaNs

In [331]:
# Valores la columna Size cuyo Outlet_Types es igual a "Supermarket Type1"
df[(df["Outlet_Type"]=="Supermarket Type1")]["Outlet_Size"].value_counts()

Small     1860
High       932
Medium     930
Name: Outlet_Size, dtype: int64

No hay correlación entre los Outlet_Types que son "Supermarket Type1" y la columna Size

In [334]:
# Valores de la columna "Outlet_Location_Type" para todos los valores NaN en "Outlet_Size"
df[df["Outlet_Size"].isna()]["Outlet_Location_Type"].value_counts()

Tier 2    1855
Name: Outlet_Location_Type, dtype: int64

Sin ambargo, todos los valores NaN en "Outlet_Size" son "Tier 2" en la columna "Outlet_Location_Type"

In [332]:
# Agrupación de la columna "Outlet_Type" = "Supermarket Type1" por Outlet_Type y Outlet_Location_Type
df[(df["Outlet_Type"]=="Supermarket Type1")].groupby(['Outlet_Type', 'Outlet_Location_Type'])["Outlet_Size"].value_counts()

Outlet_Type        Outlet_Location_Type  Outlet_Size
Supermarket Type1  Tier 1                Medium         930
                                         Small          930
                   Tier 2                Small          930
                   Tier 3                High           932
Name: Outlet_Size, dtype: int64

Vemos que para todas las filas cuyo "Outlet_Type" = "Supermarket Type1" y "Outlet_Location_Type" = "Tier 2", el "Outlet_size" es "Small"

**Estrategia:**

Para todos los outlets cuyo tamaño es NaN y que sean del tipo "Supermarket Type1" y "Tier2", usar el valor "Small" como sustituto.

In [339]:
df.loc[(df["Outlet_Type"]=="Supermarket Type1") & (df["Outlet_Location_Type"]=="Tier 2") & (df["Outlet_Size"].isna()), "Outlet_Size"] = "Small"


### Confirmen que no hay valores faltantes

In [340]:
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

### Encuentren y arreglen alguna categoría inconsistente de datos

In [341]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8519 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8519 non-null   object 
 1   Item_Weight                8519 non-null   float64
 2   Item_Fat_Content           8519 non-null   object 
 3   Item_Visibility            8519 non-null   float64
 4   Item_Type                  8519 non-null   object 
 5   Item_MRP                   8519 non-null   float64
 6   Outlet_Identifier          8519 non-null   object 
 7   Outlet_Establishment_Year  8519 non-null   int64  
 8   Outlet_Size                8519 non-null   object 
 9   Outlet_Location_Type       8519 non-null   object 
 10  Outlet_Type                8519 non-null   object 
 11  Item_Outlet_Sales          8519 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 865.2+ KB


In [342]:
df["Item_Fat_Content"].value_counts()

Low Fat    5088
Regular    2886
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [345]:
df.loc[df["Item_Fat_Content"] == "LF", "Item_Fat_Content"] = "Low Fat"
df.loc[df["Item_Fat_Content"] == "low fat", "Item_Fat_Content"] = "Low Fat"
df.loc[df["Item_Fat_Content"] == "reg", "Item_Fat_Content"] = "Regular"

df["Item_Fat_Content"].value_counts()

Low Fat    5516
Regular    3003
Name: Item_Fat_Content, dtype: int64

In [346]:
df["Item_Type"].value_counts()

Fruits and Vegetables    1232
Snack Foods              1199
Household                 910
Frozen Foods              855
Dairy                     681
Canned                    649
Baking Goods              647
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

In [352]:
df["Outlet_Size"].value_counts()

Small     4797
Medium    2790
High       932
Name: Outlet_Size, dtype: int64

In [351]:
df["Outlet_Location_Type"].value_counts()

Tier 3    3347
Tier 2    2785
Tier 1    2387
Name: Outlet_Location_Type, dtype: int64

In [350]:
df["Outlet_Type"].value_counts()

Supermarket Type1    5577
Grocery Store        1082
Supermarket Type3     932
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

###  Para cualquier columna numérica, obtengan las estadísticas resumidas para cada uno

In [353]:
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8519.000000      8519.000000  8519.000000                8519.000000   
mean     12.875420         0.066112   141.010019                1997.837892   
std       4.646098         0.051586    62.283594                   8.369105   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.785000         0.026983    93.844900                1987.000000   
50%      12.650000         0.053925   143.047000                1999.000000   
75%      16.850000         0.094558   185.676600                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8519.000000  
mean         2181.188779  
std          1706.511093  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3100.630600  
max         13086.964800